src/server/services/graph/handleSuffixUnits.js

```diff
@@ -0,0 +1,119 @@
+ * @param {*} slope The conversion's slope from source to destination.
+ * @param {*} intercept The conversion's intercept from source to destination.
+ * @param {*} unitName The new unit's name.
+ * @param {*} conn The connection to use.
+ */
+async function addNewUnitAndConversion(sourceId, destinationId, slope, intercept, unitName, graph, conn) {
+	const sourceUnit = await Unit.getById(sourceId, conn);
+	const destinationUnit = await Unit.getById(destinationId, conn);
+	const newUnit = new Unit(undefined, unitName, unitName, Unit.unitRepresentType.UNUSED, sourceUnit.secInRate,
+		Unit.unitType.SUFFIX, undefined, '', destinationUnit.displayable, destinationUnit.preferredDisplay, 'suffix unit created by OED');
+	await newUnit.insert(conn);
+
+	// Create the conversion from the prefix unit to this new unit.
+	const newConversion = new Conversion(sourceId, newUnit.id, false, slope, intercept,
+		`${sourceUnit.name} → ${newUnit.name} (created by OED for unit with prefix)`);
+	await newConversion.insert(conn);
+
+	// Add the new node and conversion to the graph.
+	graph.addNode(newUnit.id, newUnit.name);
+	graph.addLink(sourceId, newUnit.id);
+}
+
+/**
+ * Verifies that the conversion from source to destination has not changed. If so, update the conversion.
+ * @param {*} expectedSlope The expected slope.
+ * @param {*} expectedIntercept The expected intercept.
+ * @param {*} sourceId The source unit's id.
+ * @param {*} destinationId The destination unit's id.
+ * @param {*} conn The connection to use.
+ */
+async function verifyConversion(expectedSlope, expectedIntercept, sourceId, destinationId, conn) {
+	const currentConversion = await Conversion.getBySourceDestination(sourceId, destinationId, conn);
+	if (currentConversion.slope !== expectedSlope || currentConversion.intercept !== expectedIntercept) {
+		currentConversion.slope = expectedSlope;
+		currentConversion.intercept = expectedIntercept;
+		await currentConversion.update(conn);
+	}
+}
+
+/**
+ * This function is called after adding suffix units. These new units will replace
+ * the original one so we need to hide and remove all the edges from it to others.
+ * @param {*} unit The suffix unit to hide.
+ * @param {*} paths All shortest paths from this suffix unit to others.
+ * @param {*} graph The conversion graph.
+ * @param {*} conn The connection to use.
+ */
+async function hideSuffixUnit(unit, paths, graph, conn) {
+	// Hides the suffix unit since we added the units based on it if not previously done.
+	if (unit.displayable !== Unit.displayableType.NONE) {
+		unit.displayable = Unit.displayableType.NONE;
+		await unit.update(conn);
+	}
+
+	for (const p of paths) {
+		const secondUnit = await Unit.getById(p[1].id, conn);
+		// The paths to suffix units shouldn't be deleted.
+		if (secondUnit.unitType !== Unit.unitType.SUFFIX) {
+			// Removes the conversion from the conversion graph.
+			graph.removeLink(p[0].id, p[1].id);
+			// Removes the conversion from the database.
```

Comment:
If I'm correct, we rebuild the conversion graph whenever OED starts. If not, I agree that we don't need to remove the conversion from the DB.

Take the sample values in the document to be the example. If we run the handleSuffixUnits function for the first time, "metric ton of CO2" and "kg of CO2" will be added to the DB. Then, we restart OED and rebuild the graph. However, note that the new graph will contain the conversion from "kg CO2" to "kg" (since this conversion is still in the DB). Even if we run handleSuffixUnits one more time, this conversion will not be removed (since all the suffix units related to "kg CO2" have been added so the handleSuffixUnits function will do nothing). This conversion will lead to the wrong Cik table.

Reply:
The conversion graph will be rebuilt when OED starts and each time a conversion is modified/added.

I think I may be missing something because it seems to me the graph would remove the "kg CO2" to "kg" edge. Line 126 does:
await hideSuffixUnit(unit, paths, graph, conn);
for each suffix unit. Since "kg CO2" is still a suffix unit in the unit table (if you leave instead of removing), it should call this function with all paths from this suffix unit. I think there will again be three paths (to "kg", "kg of CO2" and "Metric ton of CO2"). The last two will be suffix and left alone. The first from "kg CO2" to "kg" will again be removed since it is not a suffix unit. Is this wrong?

Now for the other case I'm worried about. Suppose an admin adds the unit "lb" and a conversion from "kg" to "lb". If this had been done originally then OED would add a vertex in the graph from "kg CO2" to "lb" so we want this added now. The only way I see that happening is if there is a link from "kg CO2" to "kg" so it can then follow it to "lb". The path from "kg CO2" to "kg" already exists in the conversions so won't be added again. The new path from "kg CO2" to "lb" will be added to the conversions and also added to Cij. This is why I had the "kg CO2" to "kg" conversion staying in the conversion table but not being displayed.

Finally, removing a unit might confuse an admin as it was there before. Hiding it is easier to understand (at least from my perspective). This isn't absolutely necessary but nice.

I welcome your thoughts and thank you for helping think about these complex cases.

Reply:
> I think I may be missing something because it seems to me the graph would remove the "kg CO2" to "kg" edge. Line 126 does:
await hideSuffixUnit(unit, paths, graph, conn);
for each suffix unit. Since "kg CO2" is still a suffix unit in the unit table (if you leave instead of removing), it should call this function with all paths from this suffix unit. I think there will again be three paths (to "kg", "kg of CO2" and "Metric ton of CO2"). The last two will be suffix and left alone. The first from "kg CO2" to "kg" will again be removed since it is not a suffix unit. Is this wrong?


I've changed the get_suffix.sql a little bit so it will not return units with displayable of type NONE. So when we run the second time, the suffixUnits in line 89 will not contain "kg CO2" hence the graph still has the conversion from "kg CO2" to "kg".

The reason why I changed the get_suffix.sql is that it causes a problem if we run handleSuffixUnits for "kg CO2" one more time. Since we already had conversions from "kg CO2" to "kg of CO2" and "Metric ton of CO2" from the first time, it will create new suffix units "kg of CO2 of CO2" and "Metric ton of CO2 of CO2". After reading your comment, I think changing get_suffix.sql is not a good idea.

Alternatively, I plan to check if the destination unit in line 103 is a suffix unit. I think that we only need to run addNewUnitAndConversion if the destination is not a suffix unit. So this way will prevent creating units such as "kg of CO2 of CO2" and the unnecessary conversions will be removed as well. Please let me know what do you think about this. Thank you.

Reply:
I get it now. The pseudocode in the design document checked to avoid recreating the new suffix unit again from the same path. However, it misses that the new path gives the unit a new name. Thanks for catching this and thinking about it.

I have thought about your proposed solution and I think it is fine. If a suffix unit is the final destination of a path then there are no units it converts into. If such a vertex existed in the graph originally then no new unit/conversion would be created by the suffix process. I cannot think of a use case at this time to have such an item in the graph where it is placed by an admin. Thus, ignoring them as proposed seems fine. I say do that and don't ignore internal suffix units as the current code does.